tools/docker/Dockerfile

```diff
@@ -0,0 +1,270 @@
+# syntax=docker/dockerfile:experimental
+# Copyright 2018-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may
+# not use this file except in compliance with the License. A copy of the
+# License is located at
+#
+# 	http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+
+#########################################
+#
+# BUILD IMAGES
+#
+#########################################
+
+
+
+# Common tools needed for the build stages ahead. The final test images do not inherit directly from here, so this bloat
+# is dropped in those final end-use images.
+FROM debian:stretch as build-base
+ENV PATH="/bin:/usr/bin:/usr/local/bin:/sbin:/usr/sbin:/usr/local/sbin:/usr/lib/go/bin" \
+	DEBIAN_FRONTEND="noninteractive" \
+	GO111MODULE="on"
+RUN mkdir -p /etc/apt/sources.list.d \
+	&& echo "deb http://ftp.debian.org/debian stretch-backports main" > /etc/apt/sources.list.d/stretch-backports.list \
+	&& apt-get update \
+	&& apt-get --target-release stretch-backports install --yes --no-install-recommends \
+		golang-go \
+	&& apt-get install --yes --no-install-recommends \
+		build-essential \
+		ca-certificates \
+		curl \
+		git \
+		libdevmapper-dev \
+		libseccomp-dev \
+		musl-tools \
+		pkg-config \
+		util-linux
+
+# Run as non-root for the rest of the build stages now that we've gotten apt installs out of the way.
+RUN useradd --create-home --uid 1001 builder \
+	&& mkdir /output \
+	&& chown builder /output \
+	&& mkdir -p /home/builder/go
+
+# Buildkit caches don't support anything like a "--chown" flag yet, so ensure builder will have access to them
+RUN mkdir -p /home/builder/go/pkg/mod/cache \
+	mkdir -p /home/builder/cargo/registry \
+	&& chown -R builder /home/builder/
+
+USER builder
+WORKDIR /home/builder
+SHELL ["/bin/bash", "-c"]
+
+
+
+
+# Build firecracker itself
+FROM build-base as firecracker-build
+ENV RUSTUP_HOME="/home/builder/rustup" \
+	CARGO_HOME="/home/builder/cargo" \
+	PATH="/home/builder/cargo/bin:$PATH" \
+	RUST_VERSION="1.32.0"
+
+RUN curl --silent --show-error --retry 3 --max-time 30 --output rustup-init \
+	"https://static.rust-lang.org/rustup/archive/1.16.0/x86_64-unknown-linux-gnu/rustup-init" \
+	&& echo "2d4ddf4e53915a23dda722608ed24e5c3f29ea1688da55aa4e98765fc6223f71 rustup-init" | sha256sum -c - \
+	&& chmod +x rustup-init \
+	&& ./rustup-init -y --no-modify-path --default-toolchain $RUST_VERSION
+
+RUN --mount=type=cache,from=build-base,source=/home/builder/cargo/registry,target=/home/builder/cargo/registry \
+	source /home/builder/cargo/env \
+	&& rustup target add x86_64-unknown-linux-musl \
```

Comment:
Any particular reason the `rustup target add` is done as part of this `RUN` block and not the previous one?

Reply:
No it must have just ended up there as I was modifying other things, agree it makes more sense in the previous block, fixed.